mrc-ide/wodin
# mrc-3584 Full session rehydrate

This branch restores all session data and plots when load session from Sessions page. It re-runs model (Run tab) and sensitivity if these had previously been run. ModelFit is not re-run (we don't want to re-fit and change parameter values), but model fit is plotted from the model run solution if rehydration is detected. 

This uses the `result.inputs` values in the serialised state to rehydrate model runs from the actual inputs which were previously used, not the current values in the state (which may have changed without re-running). If compile was required when the session was last saved, the model is assumed to be stale, and is not re-run. 

To test:
- open day 1 or day 2 app and make changes e.g. code changes, upload data, fit model 
- reload the app to create a new current session
- browse to All Sessions from the Sessions menu
- click on the load icon for the second session in the list (i.e. the latest session which is not the current one)
- You should see that your session is loaded with all data, parameter values etc, and plots are rendered as expected

Key changes:
- `Rehydrate` action deserialises the loaded session state into the store, and where appropriate, recompiles, re-runs model and re-runs sensitivity to rebuild models to their state when session was last saved
- `derialise` method in `serialise` module -similar to load in HINT,  this overwrites a target AppState with serialised state values
- fetch of `OdinRunner` handled separately when rehydrating so it doesn't get removed by deserialisation
- compile, run and sensitivity actions now have `..OnRehydrate` versions which only take the basic actions required to build the solutions etc, without the additional steps which happen when not rehydrating to ensure the rest of the state is updated to accommodate the action's changes e.g. setting `compileRequired` or updating parameters
- `FitPlot` detects that session is rehydrated, so it should plot from the model run solution (by checking if model fit has a `result` and no `error` but also no `solution` - which is not serialised)

Known issue:
- open visualisation tab is not restored - this is because we just haven't implemented 2-way binding for that yet. (There's a [ticket for this]([url](https://mrc-ide.myjetbrains.com/youtrack/issue/mrc-3658)))

This branch also includes plotted data summaries in `WodinOdePlot` - this adds hidden elements with attributes containing summary values of plotted data, so that we can test the plotted data in the e2e tests without needing to rifle through plotly's svg elements. For each plotted series we create a hidden element with class `.wodin-plot-data-summary-series` and attributes:
- name
- count (number of data points)
- x-min
- x-max
- y-min
- y-max
- mode (line or markers)
- line-color (if relevant)
- marker-color (if relevant)

Big line count for this changeset, but a lot of it is moving things around, e.g. moving common code for actions for rehydrate, and common e2e test code.
